### DISCUSSION TITLE
Is it safe to modify a build instance after calling  `Complete()`?

### DISCUSSION THREAD

**Discussion by shykes** [cuelang/cue#445] [General] [answered]
I am trying to implement a custom dependency fetcher. Given a `build.Instance`, I want to detect its unresolved dependencies, and if possible, resolve them.

I found a possible solution, but would like to confirm that it is safe and uses the API correctly:

```
import (
  cueBuild "cuelang.org/go/cue/build"
)

func resolveImports(buildInst *cueBuild.Instance) (err error) {
  // Call Complete()
  err = buildInst.Complete()
  if err != nil {
    return
  }
  // Iterate on detected imports
  for _, importPath := range buildInst.ImportPaths {
    // Custom dependency fetching here
    var importInst *cueBuild.Instance
    buildInst.Imports = append(buildInst.Imports, importInst)
  }
  return
}
```

In summary:

- I call `Complete()` which populates `ImportPaths`
- For each item in `ImportPaths`, I fetch and assemble a new build instance, then add it to `Imports`

Is it safe to do this? Is there a better way?

**Comment by mpvl** [comment, marked as answer]
That should be fine. The better way would be to fix `cue.Build` so that it doesn't need an in-sync Imports anymore. All of the information is contained in `Files`.

**Reply by shykes** [reply]
Thanks!

**Comment by cueckoo** [comment]
This discussion has been migrated to https://github.com/cue-lang/cue/issues/445.

For more details about CUE's migration to a new home, please see https://github.com/cue-lang/cue/issues/1078.
